compiler/src/iree/compiler/Codegen/Common/LinalgOpInfo.cpp

```diff
@@ -91,6 +92,40 @@ static bool isTransposeLinalgOp(linalg::LinalgOp linalgOp) {
   return isInputTransposed || isOutputTransposed;
 }
 
+/// Returns true if the index map represents a transpose that benefits from
+/// shared mem.
+static bool isSharedMemTranspose(AffineMap indexMap) {
+  if (!indexMap.isEmpty() && indexMap.isPermutation()) {
+    // Ensure that the fasted moving dimension (the last one) is permuted,
+    // Otherwise shared memory promotion will not benefit the operation.
+    if (indexMap.getDimPosition(indexMap.getNumDims() - 1) !=
+        indexMap.getNumDims() - 1) {
+      return true;
+    }
+  }
```

Comment:
IIUC, we need to filter out projected permutations and transposes that do not permute the last dimension. I think those are very specific constraints that we couldn't reuse in general. However, we can extend `isTranspose` and pass a `[](AffineMap){...}` lambda that applies specific constrains to the map. In that way, the main logic would be generic and reusable but we could also extending it with specific checks. WDYT?